File: frontend/src/api/quotePdf.ts
/**
 * BUILD 15 — Quote PDF API adapter.
 *
 * Functions:
 *   generateQuotePdf(quoteId) — trigger PDF generation
 *   downloadLatestPdf(quoteId) — download most recent PDF
 *   getQuotePdfArchive(quoteId) — list all generated PDFs
 *   downloadArchivedPdf(quoteId, archiveId) — download specific version
 *
 * Rules:
 *   - No silent mock fallback.
 *   - Throws on backend error.
 *   - Uses additive /pdf/ route (does NOT touch /commercial-document).
 */

import { getAPIBaseURL } from "@/lib/config";

// ---------------------------------------------------------------------------
// Types
// ---------------------------------------------------------------------------
export interface PdfArchiveRecord {
  id: number;
  quote_id: number;
  quote_code: string;
  quote_version: number;
  filename: string;
  file_size_bytes: number | null;
  content_hash: string | null;
  generated_by: string | null;
  created_at: string | null;
}

// ---------------------------------------------------------------------------
// API Functions
// ---------------------------------------------------------------------------

/**
 * Generate a new PDF for the given quote.
 * Returns the archive record on success.
 */
export async function generateQuotePdf(
  quoteId: number
): Promise<PdfArchiveRecord> {
  const base = getAPIBaseURL();
  const url = `${base}/api/v1/entities/quotes/${quoteId}/pdf/generate`;

  const res = await fetch(url, {
    method: "POST",
    credentials: "include",
    headers: { "Content-Type": "application/json" },
  });

  if (!res.ok) {
    const detail = await res.text().catch(() => "Unknown error");
    throw new Error(
      `Failed to generate PDF for quote ${quoteId}: ${res.status} — ${detail}`
    );
  }

  return res.json();
}

/**
 * Download the most recent PDF for a quote.
 * Triggers a browser file download.
 */
export async function downloadLatestPdf(quoteId: number): Promise<void> {
  const base = getAPIBaseURL();
  const url = `${base}/api/v1/entities/quotes/${quoteId}/pdf/latest`;

  const res = await fetch(url, {
    method: "GET",
    credentials: "include",
  });

  if (!res.ok) {
    const detail = await res.text().catch(() => "Unknown error");
    throw new Error(
      `Failed to download PDF for quote ${quoteId}: ${res.status} — ${detail}`
    );
  }

  const blob = await res.blob();
  const downloadUrl = URL.createObjectURL(blob);
  const a = document.createElement("a");
  a.href = downloadUrl;

  // Extract filename from Content-Disposition header if available
  const disposition = res.headers.get("Content-Disposition");
  const filenameMatch = disposition?.match(/filename="?([^"]+)"?/);
  a.download = filenameMatch?.[1] || `oferta_${quoteId}.pdf`;

  document.body.appendChild(a);
  a.click();
  document.body.removeChild(a);
  URL.revokeObjectURL(downloadUrl);
}

/**
 * Get the list of all generated PDFs for a quote (archive history).
 */
export async function getQuotePdfArchive(
  quoteId: number
): Promise<PdfArchiveRecord[]> {
  const base = getAPIBaseURL();
  const url = `${base}/api/v1/entities/quotes/${quoteId}/pdf/archive`;

  const res = await fetch(url, {
    method: "GET",
    credentials: "include",
    headers: { "Content-Type": "application/json" },
  });

  if (!res.ok) {
    const detail = await res.text().catch(() => "Unknown error");
    throw new Error(
      `Failed to fetch PDF archive for quote ${quoteId}: ${res.status} — ${detail}`
    );
  }

  return res.json();
}

/**
 * Download a specific archived PDF by archive ID.
 * Triggers a browser file download.
 */
export async function downloadArchivedPdf(
  quoteId: number,
  archiveId: number
): Promise<void> {
  const base = getAPIBaseURL();
  const url = `${base}/api/v1/entities/quotes/${quoteId}/pdf/${archiveId}/download`;

  const res = await fetch(url, {
    method: "GET",
    credentials: "include",
  });

  if (!res.ok) {
    const detail = await res.text().catch(() => "Unknown error");
    throw new Error(
      `Failed to download archived PDF ${archiveId}: ${res.status} — ${detail}`
    );
  }

  const blob = await res.blob();
  const downloadUrl = URL.createObjectURL(blob);
  const a = document.createElement("a");
  a.href = downloadUrl;

  const disposition = res.headers.get("Content-Disposition");
  const filenameMatch = disposition?.match(/filename="?([^"]+)"?/);
  a.download = filenameMatch?.[1] || `oferta_${quoteId}_${archiveId}.pdf`;

  document.body.appendChild(a);
  a.click();
  document.body.removeChild(a);
  URL.revokeObjectURL(downloadUrl);
}

File: backend/routers/quote_pdf.py
"""
BUILD 15 — Quote PDF Generation Router.

Endpoints (ADDITIVE — does NOT modify existing commercial-document endpoints):
  POST /api/v1/entities/quotes/{quote_id}/pdf/generate
  GET  /api/v1/entities/quotes/{quote_id}/pdf/latest
  GET  /api/v1/entities/quotes/{quote_id}/pdf/archive
  GET  /api/v1/entities/quotes/{quote_id}/pdf/{archive_id}/download

All endpoints are read-only except POST /generate which creates a PDF + archive record.
No mutations to quotes, orders, inventory, or cost engine.
"""

import logging

from fastapi import APIRouter, Depends, HTTPException
from fastapi.responses import Response
from sqlalchemy.ext.asyncio import AsyncSession

from core.database import get_db
from dependencies.auth import get_current_user
from dependencies.permissions import require_permission
from services.quote_pdf_service import QuotePdfService

logger = logging.getLogger(__name__)

router = APIRouter(
    prefix="/api/v1/entities/quotes",
    tags=["quote-pdf"],
    dependencies=[Depends(get_current_user)],
)


@router.post("/{quote_id}/pdf/generate", status_code=201)
async def generate_quote_pdf(
    quote_id: int,
    db: AsyncSession = Depends(get_db),
    current_user: dict = Depends(require_permission("quote.export_pdf")),
):
    """Generate a new PDF from the current quote state.

    Creates a PDF file and stores an archive record with traceability.
    Does NOT modify the quote or any other entity.
    """
    service = QuotePdfService(db)
    generated_by = current_user.get("email") if isinstance(current_user, dict) else None

    result = await service.generate_pdf(quote_id, generated_by=generated_by)

    if "error" in result:
        error = result["error"]
        if error == "quote_not_found":
            raise HTTPException(status_code=404, detail="Quote not found")
        if error == "pdf_generation_failed":
            raise HTTPException(status_code=500, detail="PDF generation failed")
        raise HTTPException(status_code=500, detail=error)

    return result


@router.get("/{quote_id}/pdf/latest")
async def get_latest_pdf(
    quote_id: int,
    db: AsyncSession = Depends(get_db),
):
    """Download the most recent generated PDF for a quote.

    Returns PDF bytes with Content-Disposition attachment header.
    """
    service = QuotePdfService(db)
    archive = await service.get_latest(quote_id)

    if not archive:
        raise HTTPException(
            status_code=404,
            detail="No PDF generated for this quote",
        )

    pdf_bytes = service.get_pdf_bytes(archive)
    if not pdf_bytes:
        raise HTTPException(
            status_code=404,
            detail="PDF file not found on disk",
        )

    return Response(
        content=pdf_bytes,
        media_type="application/pdf",
        headers={
            "Content-Disposition": f'attachment; filename="{archive.filename}"',
        },
    )


@router.get("/{quote_id}/pdf/archive")
async def get_pdf_archive(
    quote_id: int,
    db: AsyncSession = Depends(get_db),
):
    """List all generated PDF documents for a quote (history).

    Returns archive records ordered by newest first.
    """
    service = QuotePdfService(db)
    archives = await service.get_archive_list(quote_id)

    return [
        {
            "id": a.id,
            "quote_id": a.quote_id,
            "quote_code": a.quote_code,
            "quote_version": a.quote_version,
            "filename": a.filename,
            "file_size_bytes": a.file_size_bytes,
            "content_hash": a.content_hash,
            "generated_by": a.generated_by,
            "created_at": a.created_at.isoformat() if a.created_at else None,
        }
        for a in archives
    ]


@router.get("/{quote_id}/pdf/{archive_id}/download")
async def download_archived_pdf(
    quote_id: int,
    archive_id: int,
    db: AsyncSession = Depends(get_db),
):
    """Download a specific archived PDF by its archive ID.

    Validates that the archive belongs to the specified quote_id.
    """
    service = QuotePdfService(db)
    archive = await service.get_archive_by_id(quote_id, archive_id)

    if not archive:
        raise HTTPException(
            status_code=404,
            detail="Archive record not found or does not belong to this quote",
        )

    pdf_bytes = service.get_pdf_bytes(archive)
    if not pdf_bytes:
        raise HTTPException(
            status_code=404,
            detail="PDF file not found on disk",
        )

    return Response(
        content=pdf_bytes,
        media_type="application/pdf",
        headers={
            "Content-Disposition": f'attachment; filename="{archive.filename}"',
        },
    )